File: leetcode-algorithms/019. Remove Nth Node From End of List/solution.cpp
#include <iostream>
using namespace std;
ListNode* removeNthFromEnd(ListNode* head, int n) {
	ListNode f(0);
	f.next = head;
	head = &f;

	if(head==NULL || n < 0)
		return head;
	ListNode* p1,*p2;
	p1 = p2 = head;
	for(int i=0;i<n;i++){
		if(p2==NULL)
			return NULL;
		p2 = p2->next;
	}
	while(p2->next){
		p2 = p2->next;
		p1 = p1->next;
	}
	p1->next = p1->next->next;
	return head->next;
}
int main()
{
	return 0;
}

File: leetcode-algorithms/019. Remove Nth Node From End of List/solution.py
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None


def removeNthFromEnd(head, n):
	"""
	:type head: ListNode
	:type n: int
	:rtype: ListNode
	"""

	fast = slow = head
	for _ in range(n):
		fast = fast.next
	if not fast:
		return head.next
	while fast.next:
		fast = fast.next
		slow = slow.next
	slow.next = slow.next.next
	return head




if __name__ == '__main__':
	a = ListNode(1)
	b = ListNode(2)
	a.next = b
	c = ListNode(3)
	b.next = c
	d = ListNode(4)
	c.next = d
	e = ListNode(5)
	d.next = e

	head = removeNthFromEnd(a,2)

	while head:
		print(head.val)
		head = head.next



